From d0f29f56056e5205382937f190cc74e0962f599f Mon Sep 17 00:00:00 2001
From: kszaq <kszaquitto@gmail.com>
Date: Thu, 3 Mar 2016 07:43:23 +0100
Subject: [PATCH] Save settings only if they were modified after the last save

This prevents from multiple saving the same settings and helps to resolve
the issue on Amlogic G18REF TV-boxes when setiings may be lost after a poweroff.

On G18REF When you press the red button on the remote the system receives a signal that power
is about to off. XBMC always writes guisettings.xml before exit, and the same settings
may be written several times from different places in code. But the power gets turned off
before the system completes all shutdown procedures. There may be the case that guisettings.xml
is written half-way and couldn't be read upon next boot, so the XBMC creates a new one with
default settings.

With this fix the settings will be written at exit only once, minimizing the risk of being lost.

Original patch by Alex Deryskyba <alex@codesnake.com>
Modified by kszaq: apply to all XML files
---
 xbmc/utils/XBMCTinyXML.cpp | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/xbmc/utils/XBMCTinyXML.cpp b/xbmc/utils/XBMCTinyXML.cpp
index c9c7b49..ad3f94a 100644
--- a/xbmc/utils/XBMCTinyXML.cpp
+++ b/xbmc/utils/XBMCTinyXML.cpp
@@ -26,7 +26,9 @@
 #include "utils/Utf8Utils.h"
 #include "LangInfo.h"
 #include "RegExp.h"
+#include "Util.h"
 #include "utils/log.h"
+#include "utils/md5.h"
 
 #define MAX_ENTITY_LENGTH 8 // size of largest entity "&#xNNNN;"
 #define BUFFER_SIZE 4096
@@ -118,6 +120,15 @@ bool CXBMCTinyXML::SaveFile(const char *_filename) const
 bool CXBMCTinyXML::SaveFile(const std::string& filename) const
 {
   XFILE::CFile file;
+  if (file.Exists(filename))
+  {
+    std::string fileMD5 = CUtil::GetFileMD5(filename);
+    TiXmlPrinter xmlPrinter;
+    Accept(&xmlPrinter);
+    std::string settingsMD5 = XBMC::XBMC_MD5::GetMD5(xmlPrinter.CStr());
+    if (fileMD5 == settingsMD5)
+      return true;
+  }
   if (file.OpenForWrite(filename, true))
   {
     TiXmlPrinter printer;
-- 
1.9.1

